projects/seeker/agents/seeker.py

```diff
@@ -0,0 +1,905 @@
+        """
+        Full batch_act pipeline.
+
+        :param observations:
+            batchsize-length list of observations from self.observe
+
+        :return reply:
+            return batchsize-length list of final replies.
+        """
+        knowledge_agent_observations = [o['knowledge_agent'] for o in observations]
+        # First, determine whether we're searching
+        batch_reply_sdm, search_indices, knowledge_agent_observations = self.batch_act_sdm(
+            observations, knowledge_agent_observations
+        )
+        # Second, generate search queries
+        batch_reply_sqm = self.batch_act_sqm(observations, search_indices)
+
+        # Third, generate the knowledge sentence
+        batch_reply_krm = self.batch_act_krm(
+            observations, knowledge_agent_observations, search_indices
+        )
+
+        # Fourth, generate the dialogue response!
+        batch_reply_drm = self.batch_act_drm(observations, batch_reply_krm)
+
+        # Finaly, combine them all in the drm batch reply.
+        for sdm, sqm, krm, drm in zip(
+            batch_reply_sdm, batch_reply_sqm, batch_reply_krm, batch_reply_drm
+        ):
+            if drm.is_padding():
+                continue
+            drm.force_set('search_decision', sdm.get('text', ''))
+            drm.force_set('search_query', sqm.get('text', ''))
+            drm.force_set('knowledge_response', krm.get('text', ''))
+            docs = krm.get('top_docs', [Document("", "", "")])
+            drm.force_set('doc_titles', [d.get_title() for d in docs])
+            drm.force_set('doc_content', [d.get_text() for d in docs])
+            drm.force_set('doc_urls', [d.get_id() for d in docs])
+
+        return batch_reply_drm
+
+    def act(self):
+        """
+        Call batch_act with the singleton batch.
+        """
+        response = self.batch_act([self.observations])[0]
+        self.self_observe(response)
+        return response
+
+    def self_observe(self, self_message: Message):
+        """
+        Override TA.self_observe.
+
+        Make sure that knowledge agent and other agents have the same history.
+
+        This eliminates unnecessary copies of the previous knowledge in the history.
+        """
+        self.knowledge_agent.self_observe(self_message)
+        self.knowledge_responses.append(self_message.get('knowledge_response', ''))
+        observation = {'text': self.knowledge_agent.history.get_history_str()}
+        for agent in self.dialogue_agent_clones:
+            agent.reset()
+            agent.history.update_history(
+                observation,
+                temp_history=self.dialogue_agent.get_temp_history(observation),
+            )
+        if (
+            self.search_decision_agent
+            and self.search_decision is SearchDecision.COMPUTE
+        ):
+            self.search_decision_agent.self_observe(self_message)
+            self.search_decision_agent.history.reset()
```

Comment:
Does the decision agent need to self_observe with the history getting reset right after?

Reply:
i'm calling `history.reset` directly, so the agent still needs a self observe.

This may be old code though; what i've realized is now I can just call `self.search_decision_agent.reset()`, which will reset the invariants